File: Example-OC/ViewController.m
//
//  ViewController.m
//  Example-OC
//
//  Created by RayJiang on 2019/7/18.
//  Copyright © 2019 RayJiang. All rights reserved.
//

#import "ViewController.h"
#import "SceneDelegate.h"
#import "XYColorOC/XYColorOC.h"

@interface ViewController ()
@property (nonatomic, strong) UIColor *color1;
@property (nonatomic, strong) UIButton *button;
@property (nonatomic, strong) UILabel *label;
@property (nonatomic, strong) CALayer *layer1;
@property (nonatomic, strong) CALayer *layer2;
@property (nonatomic, strong) CALayer *layer3;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.view.backgroundColor = UIColor.systemBackgroundColor;
    [self.view addSubview:self.button];
    [self.view addSubview:self.label];
    [self.view.layer addSublayer:self.layer1];
    [self.view.layer addSublayer:self.layer2];
    [self.view.layer addSublayer:self.layer3];
}

- (void)buttonTapped:(UIButton *)sender {
    for (UIScene *scene in UIApplication.sharedApplication.connectedScenes) {
        SceneDelegate *delegate = (SceneDelegate *)scene.delegate;
        delegate.window.overrideUserInterfaceStyle = self.view.traitCollection.userInterfaceStyle == UIUserInterfaceStyleDark ? UIUserInterfaceStyleLight : UIUserInterfaceStyleDark;
    }
    
}

- (UIColor *)color1 {
    if (!_color1) {
        _color1 = [UIColor colorWithDynamicProvider:^UIColor * _Nonnull(UITraitCollection * _Nonnull traitCollection) {
            if (traitCollection.userInterfaceStyle == UIUserInterfaceStyleDark) {
                return UIColor.yellowColor;
            } else {
                return UIColor.greenColor;
            }
        }];
    }
    return _color1;
}

- (UIButton *)button {
    if (!_button) {
        _button = [UIButton buttonWithType:UIButtonTypeCustom];
        _button.frame = CGRectMake(UIScreen.mainScreen.bounds.size.width/2-100, 100, 200, 40);
        [_button setTitle:@"Change mode" forState:UIControlStateNormal];
        [_button setTitleColor:UIColor.labelColor forState:UIControlStateNormal];
        _button.layer.cornerRadius = 4;
        _button.layer.borderWidth = 1;
        [_button xy_setLayerBorderColor:UIColor.labelColor];
        [_button addTarget:self action:@selector(buttonTapped:) forControlEvents:UIControlEventTouchUpInside];
    }
    return _button;
}

- (UILabel *)label {
    if (!_label) {
        _label = [UILabel new];
        _label.frame = CGRectMake(UIScreen.mainScreen.bounds.size.width/2-50, 160, 100, 40);
        _label.text = @"Text";
        _label.textAlignment = NSTextAlignmentCenter;
        _label.textColor = UIColor.labelColor;
        _label.layer.cornerRadius = 4;
        _label.layer.borderWidth = 1;
        [_label xy_setLayerBorderColor:self.color1];
    }
    return _label;
}

- (CALayer *)layer1 {
    if (!_layer1) {
        _layer1 = [CALayer new];
        _layer1.frame = CGRectMake(15, 230, 80, 80);
        _layer1.cornerRadius = 4;
        [_layer1 xy_setLayerBackgroundColor:self.color1 with:self.view];
    }
    return _layer1;
}

- (CALayer *)layer2 {
    if (!_layer2) {
        _layer2 = [CALayer new];
        _layer2.frame = CGRectMake(UIScreen.mainScreen.bounds.size.width/2-40, 230, 80, 80);
        _layer2.cornerRadius = 4;
        _layer2.borderWidth = 10;
        [_layer2 xy_setLayerBackgroundColor:self.color1 with:self.view];
        [_layer2 xy_setLayerBorderColor:UIColor.labelColor with:self.view];
    }
    return _layer2;
}

- (CALayer *)layer3 {
    if (!_layer3) {
        _layer3 = [CALayer new];
        _layer3.frame = CGRectMake(UIScreen.mainScreen.bounds.size.width-80-15, 230, 80, 80);
        _layer3.cornerRadius = 4;
        _layer3.shadowRadius = 20;
        _layer3.shadowOpacity = 1;
        [_layer3 xy_setLayerShadowColor:self.color1 with:self.view];
        [_layer3 xy_setLayerBackgroundColor:UIColor.labelColor with:self.view];
    }
    return _layer3;
}

@end

File: Example/ViewController.swift
//
//  ViewController.swift
//  Example
//
//  Created by RayJiang on 2019/7/18.
//  Copyright © 2019 RayJiang. All rights reserved.
//

import UIKit
import XYColor

class ViewController: UIViewController {
    
    private lazy var color1: UIColor = {
        return UIColor { (traitCollection) -> UIColor in
            if traitCollection.userInterfaceStyle == .dark {
                return .yellow
            } else {
                return .green
            }
        }
    }()
    
    private lazy var button: UIButton = {
        let view = UIButton(type: .custom)
        view.frame = CGRect(x: UIScreen.main.bounds.width/2-100, y: 100, width: 200, height: 40)
        view.setTitle("Change mode", for: .normal)
        view.setTitleColor(UIColor.label, for: .normal)
        view.layer.cornerRadius = 4
        view.layer.borderWidth = 1
        view.setLayerBorderColor(UIColor.label)
        view.addTarget(self, action: #selector(buttonTapped(_:)), for: .touchUpInside)
        return view
    }()
    
    private lazy var label: UILabel = {
        let view = UILabel()
        view.frame = CGRect(x: UIScreen.main.bounds.width/2-50, y: 160, width: 100, height: 40)
        view.text = "Text"
        view.textAlignment = .center
        view.textColor = UIColor.label
        view.layer.cornerRadius = 4
        view.layer.borderWidth = 1
        view.setLayerBorderColor(color1)
        return view
    }()
    
    private lazy var layer1: CALayer = {
        let layer = CALayer()
        layer.frame = CGRect(x: 15, y: 230, width: 80, height: 80)
        layer.cornerRadius = 4
        layer.setBackgroundColor(color1, with: view)
        return layer
    }()
    
    private lazy var layer2: CALayer = {
        let layer = CALayer()
        layer.frame = CGRect(x: UIScreen.main.bounds.width/2-40, y: 230, width: 80, height: 80)
        layer.cornerRadius = 4
        layer.borderWidth = 10
        layer.setBackgroundColor(color1, with: view)
        layer.setBorderColor(UIColor.label, with: view)
        return layer
    }()
    
    private lazy var layer3: CALayer = {
        let layer = CALayer()
        layer.frame = CGRect(x: UIScreen.main.bounds.width-80-15, y: 230, width: 80, height: 80)
        layer.cornerRadius = 4
        layer.shadowRadius = 20
        layer.shadowOpacity = 1
        layer.setShadowColor(color1, with: view)
        layer.setBackgroundColor(UIColor.label, with: view)
        return layer
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor.systemBackground
        
        view.addSubview(button)
        view.addSubview(label)
        view.layer.addSublayer(layer1)
        view.layer.addSublayer(layer2)
        view.layer.addSublayer(layer3)
    }

    @objc private func buttonTapped(_ sender: UIButton) {
        guard let scene = UIApplication.shared.connectedScenes.first?.delegate as? SceneDelegate else { return }
        scene.window?.overrideUserInterfaceStyle = view.traitCollection.userInterfaceStyle == .dark ? .light : .dark
    }
}

